src/trace/trigger.ts

```diff
@@ -0,0 +1,270 @@
+  if (source === "dynamodb") {
+    eventSourceARN = extractDynamoDBStreamEventARN(event);
+  }
+
+  // arn:aws:kinesis:EXAMPLE
+  if (source === "kinesis") {
+    eventSourceARN = extractKinesisStreamEventARN(event);
+  }
+  return eventSourceARN;
+}
+
+/**
+ * extractHTTPTags extracts HTTP facet tags from the triggering event
+ */
+function extractHTTPTags(event: APIGatewayEvent | APIGatewayProxyEventV2 | ALBEvent) {
+  const httpTags: { [key: string]: string } = {};
+
+  if (eventType.isAPIGatewayEvent(event)) {
+    const requestContext = event.requestContext;
+    if (requestContext.domainName) {
+      httpTags["http.url"] = requestContext.domainName;
+    }
+    httpTags["http.url_details.path"] = requestContext.path;
+    httpTags["http.method"] = requestContext.httpMethod;
+    if (event.headers.Referer) {
+      httpTags["http.referer"] = event.headers.Referer;
+    }
+  }
+
+  if (eventType.isAPIGatewayEventV2(event)) {
+    const requestContext = event.requestContext;
+    httpTags["http.url"] = requestContext.domainName;
+    httpTags["http.url_details.path"] = requestContext.http.path;
+    httpTags["http.method"] = requestContext.http.method;
+    if (event.headers.Referer) {
+      httpTags["http.referer"] = event.headers.Referer;
+    }
+  }
+
+  if (eventType.isALBEvent(event)) {
+    httpTags["http.url_details.path"] = event.path;
+    httpTags["http.method"] = event.httpMethod;
+    if (event.headers && event.headers.Referer) {
+      httpTags["http.referer"] = event.headers.Referer;
+    }
+  }
+  return httpTags;
+}
+
+/**
+ * extractTriggerTags parses the trigger event object for tags to be added to the span metadata
+ */
+export function extractTriggerTags(event: any, context: Context) {
+  let triggerTags: { [key: string]: string } = {};
+  const eventSource = parseEventSource(event);
+  if (eventSource) {
+    triggerTags["function_trigger.event_source"] = eventSource;
+
+    const eventSourceARN = parseEventSourceARN(eventSource, event, context);
+    if (eventSourceARN) {
+      triggerTags["function_trigger.event_source_arn"] = eventSourceARN;
+    }
+  }
+
+  if (isHTTPTriggerEvent(eventSource)) {
+    triggerTags = { ...triggerTags, ...extractHTTPTags(event) };
+  }
+  return triggerTags;
+}
+
+/**
+ * extractHTTPStatusCode extracts a status code from the response if the Lambda was triggered
+ * by API Gateway or ALB
+ */
+export function extractHTTPStatusCodeTag(triggerTags: { [key: string]: string }, result: any) {
+  let eventSource: string | undefined;
+  triggerTags ? (eventSource = triggerTags["function_trigger.event_source"]) : (eventSource = undefined);
+  if (!isHTTPTriggerEvent(eventSource)) {
+    return;
+  }
+
+  let statusCode = "200";
+  // Return a 502 status if no response is found
+  if (result === undefined) {
+    statusCode = "502";
+  } else if (result.statusCode) {
+    // Use the statusCode if available
```

Comment:
Using any for the result type is fine for this method, but we should be stricter with the type checking, especially since this method is called during a finally block and should never throw.

```typescript
const statusCode = result?.statusCode; // Will be undefined if statusCode doesn't exist or result isn't an object

if (typeof statusCode === 'number') {
   return statusCode.toString();
} else if (typeof statusCode ==="string") {
   return statusCode;
}
```

Reply:
TIL about optional chaining! I kept the check for `result === undefined` here since we would still need to return a `502` status when the Lambda doesn't receive a response.